src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -117,11 +124,61 @@ public Http2Connection(HttpConnectionPool pool, Stream stream)
             _maxConcurrentStreams = int.MaxValue;
             _pendingWindowUpdate = 0;
 
+            if (_pool.EnableMultipleHttp2Connections)
+            {
+                _freeStreamSlots = InitialSlotsLimit;
+            }
+
             if (NetEventSource.IsEnabled) TraceConnection(stream);
         }
 
         private object SyncObject => _httpStreams;
 
+        public bool AcquireStreamSlot()
+        {
+            while (true)
+            {
+                int currentSlots = _freeStreamSlots;
+                if (currentSlots == 0)
+                {
+                    long currentExpiration = _freeStreamSlotsExpiration;
+                    if (currentExpiration > Environment.TickCount64)
+                    {
+                        // No empty slots and it's too soon to request the latest value from the credit manager.
+                        return false;
+                    }
+
+                    _freeStreamSlotsExpiration = Environment.TickCount64 + SlotExpirationPeriod;
```

Comment:
Same visibility issue exists with `_freeStreamSlotsExpiration`; making the variable `volatile` should fix it.

Reply:
Fixed with Interlocked.Read since volatile read of `long` fields is not atomic on 32-bit platforms.